apps/api/src/billing/services/wallet-manager/WalletManager.ts

```diff
@@ -0,0 +1,130 @@
+
+  private readonly HD_PATH = "m/44'/118'/0'/0";
+
+  private masterWallet: DirectSecp256k1HdWallet;
+
+  private client: SigningStargateClient;
+
+  private readonly logger = new ContextualLogger({ context: WalletManager.name });
+
+  constructor(
+    @InjectBillingConfig() private readonly config: BillingConfig,
+    private readonly rpcMessageService: RpcMessageService
+  ) {}
+
+  async createAndAuthorizeTrialSpending({ addressIndex }: { addressIndex: number }) {
+    const { address } = await this.createWallet({ addressIndex });
+    await this.authorizeSpending({
+      address,
+      limits: {
+        deployment: this.config.TRIAL_DEPLOYMENT_ALLOWANCE_AMOUNT,
+        fees: this.config.TRIAL_FEES_ALLOWANCE_AMOUNT
+      },
+      expiration: add(new Date(), { days: this.config.TRIAL_ALLOWANCE_EXPIRATION_DAYS })
+    });
+
+    return { address };
+  }
+
+  async createWallet({ addressIndex }: { addressIndex: number }) {
+    const hdPath = stringToPath(`${this.HD_PATH}/${addressIndex}`);
+    const wallet = await DirectSecp256k1HdWallet.fromMnemonic(this.config.MASTER_WALLET_MNEMONIC, {
+      prefix: this.PREFIX,
+      hdPaths: [hdPath]
+    });
+    const [account] = await wallet.getAccounts();
+    this.logger.debug({ event: "WALLET_CREATED", address: account.address });
+
+    return { address: account.address };
+  }
+
+  async authorizeSpending(options: SpendingAuthorizationOptions) {
+    try {
+      const masterWalletAddress = await this.getMasterWalletAddress();
+      const messageParams = {
+        granter: masterWalletAddress,
+        grantee: options.address,
+        denom: this.config.TRIAL_ALLOWANCE_DENOM,
+        expiration: options.expiration
+      };
+      const messages = [
+        this.rpcMessageService.getGrantMsg({
+          ...messageParams,
+          limit: options.limits.deployment
+        }),
+        this.rpcMessageService.getGrantBasicAllowanceMsg({
+          ...messageParams,
+          limit: options.limits.fees
+        })
+      ];
+      const client = await this.getClient();
+      const fee = await this.estimateFee(messages, this.config.TRIAL_ALLOWANCE_DENOM);
+      await client.signAndBroadcast(masterWalletAddress, messages, fee);
+      this.logger.debug({ event: "SPENDING_AUTHORIZED", address: options.address });
+    } catch (error) {
+      if (error.message.includes("fee allowance already exists")) {
+        this.logger.debug({ event: "SPENDING_ALREADY_AUTHORIZED", address: options.address });
+      } else {
+        error.message = `Failed to authorize spending for address ${options.address}: ${error.message}`;
+        this.logger.error(error);
+        throw error;
+      }
+    }
+  }
+
+  private async estimateFee(messages: readonly EncodeObject[], denom: string) {
+    const client = await this.getClient();
+    const address = await this.getMasterWalletAddress();
+    const gasEstimation = await client.simulate(address, messages, "allowance grant");
```

Comment:
We can remove the memo

Reply:
I think the address should be based on the current tx and not hardcoded to be the master wallet

Reply:
I thought it's a transaction on the master wallet, so the estimation of the fee should be also based on master wallet. Could you please elaborate on this comment a bit more because I don't not follow.

Reply:
> We can remove the memo

Memo is a required parameter. Can not omit it due to type